pkg/virt-controller/watch/dra/dra.go

```diff
@@ -0,0 +1,611 @@
+		// This could be an attachment pod, look up the pod, and check if it is owned by a VMI.
+		obj, exists, err := c.podIndexer.GetByKey(namespace + "/" + controllerRef.Name)
+		if err != nil {
+			return nil
+		}
+		if !exists {
+			return nil
+		}
+		pod, _ := obj.(*k8sv1.Pod)
+		controllerRef = controller.GetControllerOf(pod)
+	}
+	// We can't look up by UID, so look up by Name and then verify UID.
+	// Don't even try to look up by Name if it is nil or the wrong Kind.
+	if controllerRef == nil || controllerRef.Kind != virtv1.VirtualMachineInstanceGroupVersionKind.Kind {
+		return nil
+	}
+	vmi, exists, err := c.vmiIndexer.GetByKey(namespace + "/" + controllerRef.Name)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		return nil
+	}
+
+	if vmi.(*virtv1.VirtualMachineInstance).UID != controllerRef.UID {
+		// The controller we found with this Name is not the same one that the
+		// ControllerRef points to.
+		return nil
+	}
+	return vmi.(*virtv1.VirtualMachineInstance)
+}
+
+func (c *DRAStatusController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+	log.Log.Info("Starting DRA Status controller")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.hasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping DRA Status controller")
+}
+
+func (c *DRAStatusController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+var draStatusControllerWorkQueueTracer = &traceUtils.Tracer{Threshold: time.Second}
+
+func (c *DRAStatusController) Execute() bool {
+	if !c.clusterConfig.GPUsWithDRAGateEnabled() {
+		return false
+	}
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+
+	draStatusControllerWorkQueueTracer.StartTrace(key, "dra-status-controller VMI workqueue", trace.Field{Key: "Workqueue Key", Value: key})
+	defer draStatusControllerWorkQueueTracer.StopTrace(key)
+
+	defer c.Queue.Done(key)
+	err := c.execute(key)
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VirtualMachineInstance %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VirtualMachineInstance %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *DRAStatusController) execute(key string) error {
+	obj, exists, err := c.vmiIndexer.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return nil
+	}
+	vmi := obj.(*virtv1.VirtualMachineInstance)
+	if vmi == nil {
+		return fmt.Errorf("nil vmi reference")
+	}
+	logger := log.Log.Object(vmi)
+
+	if vmi.DeletionTimestamp != nil {
+		// object is being deleted, do not process it
+		log.Log.Info("vmi being deleted, dra status controller skipping")
+		return nil
+	}
+	// Only consider pods which belong to this vmi
+	// excluding unfinalized migration targets from this list.
+	pod, err := controller.CurrentVMIPod(vmi, c.podIndexer)
```

Comment:
I wonder why this should be an error? I think the result here will re-enqueue with rate limit, but couldn't we add the vmi on pod add or update?

Reply:
My understanding is that it is needed for transient errors where indexer is not able to return the pod list. Pod add or update events will be hit or a miss, this error needs rate-limited retries until it goes away

Reply:
I'm not sure, tbh, we are relying on pod events in other cases. I don't know if we really need to spin the controller while the pod is in scheduling and such.
I don't object this staying as it is for alpha, but I'd evaluate it later.
@xpivarc fyi, if you have an opinion.

Reply:
> I think that we can AddRateLimited() to the queue but return nil here. WDYT?

@iholder101 By default when the `c.updateStatus(...)` call returns with error, it is added to the queue with `AddRateLimited` in c.Execute(). Considering that we can keep this code as is, unless @xpivarc as other strong opinions.

Reply:
> @iholder101 By default when the `c.updateStatus(...)` call returns with error, it is added to the queue with `AddRateLimited` in c.Execute(). Considering that we can keep this code as is, unless @xpivarc as other strong opinions.

Yeah, you're absolutely right. I think @vladikr's point is that while it will invoke `AddRateLimited` which is intended, raising an error is not intended as nothing erroneous had happened.

As I see it, we have 3 options:
1. Keep your code.
2. Avoid returning an error, but explicitly requeuing.
3. Avoid returning and error & requeuing, rely on informer event handlers.

@xpivarc to give his opinion. I don't have a strong opinion about this.

Reply:
can we handle this as a follow-up? I will file an issue and make it alpha blocker.

 I am not very convinced it is safe, I would like to write a test, if possible, to help validate if 3 is a safe option.

Reply:
The e2e tests are not part of alpha feature, yet, which is why I want to take this as follow up.

Reply:
> can we handle this as a follow-up? I will file an issue and make it alpha blocker.

Unlike Kubernetes, we don't have a "test freeze". Code freeze is for both production and test code.

That being said, I think the proposed path forward makes sense.
We can frame it as a bug, fix it moving forward, and perhaps decide to backport it.
Please kindly open an issue to track that.